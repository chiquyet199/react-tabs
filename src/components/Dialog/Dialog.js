import React, {Component} from "react";
import onClickOutside from "react-onclickoutside";
import {
  DialogContainer,
  DialogButtonBar,
  DialogParagraph,
  DialogTitle,
} from "./Dialog.styles";
import Button from "../Button";

export class Dialog extends Component {
  handleClickOutside = () => {
    this.props.onClose();
  };

  render() {
    const {
      title,
      paragraph,
      confirmButtonType,
      cancelButtonText,
      onDialogCancel,
      onDialogConfirm,
      confirmButtonText,
      globaltheme,
      theme,
      fullWidthText,
      children,
    } = this.props;

    return (
      <DialogContainer theme={theme}>
        {title && (
          <DialogTitle globaltheme={globaltheme} theme={theme}>
            {title}
          </DialogTitle>
        )}
        {children}
        {paragraph && (
          <DialogParagraph
            globaltheme={globaltheme}
            theme={theme}
            fullWidthText={fullWidthText}
          >
            {paragraph}
          </DialogParagraph>
        )}
        <DialogButtonBar>
          <Button
            type="secondary"
            middle={cancelButtonText}
            onClick={onDialogCancel}
          />
          <Button
            type={confirmButtonType || "primary"}
            middle={confirmButtonText}
            onClick={onDialogConfirm}
          />
        </DialogButtonBar>
      </DialogContainer>
    );
  }
}

export default onClickOutside(Dialog);
