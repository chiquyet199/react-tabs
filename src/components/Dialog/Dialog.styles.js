import styled from "styled-components/native";
import {responsive, isWeb} from "../../utils/common";
import {TextParagraph, PageTitle, ButtonBar} from "../common/layout";

export const DialogOverlay = styled.View`
  align-items: center;
  justify-content: center;
  position: ${isWeb ? "fixed" : "absolute"};
  height: 100%;
  top: ${isWeb ? 0 : "auto"};
  right: 0;
  left: 0;
  bottom: 0;
  padding: 16px;
  background-color: ${({theme}) => theme.colors.darkTransparent};
  z-index: 9999;
  opacity: 1;
  ${isWeb &&
    `overflow-x: hidden;
     overflow-y: auto;
  `};
`;

export const DialogContainer = styled.View`
  flex-direction: column;
  max-width: 400px;
  width: 100%;
  background-color: ${({theme}) => theme.colors.white};
  border-radius: 6px;
  min-height: 100px;
  align-items: center;
  padding: 30px 25px 20px;
`;

export const DialogButtonBar = responsive(styled(ButtonBar)`
  margin: 0;
  width: 100%;
`)({
  S: `
  flex-direction: column;
  `,
});

export const DialogParagraph = styled(TextParagraph)`
  width: 100%;
  max-width: ${({fullWidthText}) => (fullWidthText ? "100%" : "80%")}
  font-size: 18px;
  padding: 0 0 15px;
  color: ${({theme}) => theme.colors.canopySteel};
`;

export const DialogTitle = styled(PageTitle)`
  text-align: center;
  width: 100%;
  padding: 0 0 20px;
  color: ${({theme}) => theme.colors.canopySteel};
`;
