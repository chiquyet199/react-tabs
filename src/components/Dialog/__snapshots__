/Users/edward/Desktop/canopy-frontend/packages/canopy-frontend-common/src/components/Dialog/__native_snapshots__