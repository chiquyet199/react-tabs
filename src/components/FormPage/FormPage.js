import React, {Component, Fragment} from "react";
import {connect} from "react-redux";
import DynamicForm from "../DynamicForm";
import Button from "../Button";
import {
  PageTitle,
  PageContainer,
  TextParagraph,
  ErrorMessage,
  ScrollContentContainer,
  ScrollButtonBar,
  ScrollScrollView,
} from "../common/layout";
import {SubmissionError} from "redux-form";
import {isEmpty} from "underscore";
import {validate} from "../../utils/validators/common";
import {isWeb} from "../../utils/common";
import {object, bool, string, shape} from "prop-types";

export class BaseFormPage extends Component {
  state = {
    displayMultiSwitchError: false,
  };

  withValidation = parentSubmit => async values => {
    const {schema, userType = "admin", inviteUser, translations} = this.props;
    const errors = await validate(schema.properties, values);
    if (!isEmpty(errors)) {
      /*
        `branch_` indicates multiple switches
        check if user selected at least one of them
      */
      if (Object.keys(errors).some(item => item.includes("branch_"))) {
        await this.setState({
          displayMultiSwitchError: true,
        });
      }

      Object.keys(errors).forEach(error => {
        errors[error] = translations.validationErrors[errors[error]];
      });

      throw new SubmissionError(errors);
    }

    if (schema.customInviteUserValidation && userType !== "user") {
      await inviteUser(values);
    }

    return parentSubmit(values);
  };

  renderHeaderAndInfo(schema) {
    const headerAndInfo = [];

    if (schema.header) {
      headerAndInfo.push(
        <PageTitle
          left="true"
          key={`page_header_${schema.header.toLowerCase().replace(" ", "_")}`}
        >
          {schema.header}
        </PageTitle>,
      );
    }
    if (schema.info) {
      headerAndInfo.push(
        <TextParagraph
          left="true"
          key={`page_info_${schema.info.toLowerCase().replace(" ", "_")}`}
        >
          {schema.info}
        </TextParagraph>,
      );
    }

    return headerAndInfo;
  }

  componentDidUpdate(prevProps) {
    if (prevProps.hasErrors !== this.props.hasErrors && this.props.hasErrors) {
      this.props.submit();
    }
  }

  render() {
    const {displayMultiSwitchError} = this.state;
    const {
      schema,
      components,
      omitSubmit,
      submit,
      onSubmit,
      submitOnEnter,
      screenSize,
      multiSwitchErrorText,
      extraButtons,
      extraButtonBarButtons,
    } = this.props;

    if (schema) {
      const initialValues = this.props.initialValues || schema.initialValues;
      const {
        buttonStyles,
        buttonIcon = "arrow-right",
        buttonText = "Next",
        buttonType = "primary",
        layout,
        marginTop,
      } = schema;

      let multiSwitchError = null;

      if (displayMultiSwitchError) {
        multiSwitchError = <ErrorMessage>{multiSwitchErrorText}</ErrorMessage>;
      }

      return (
        <ScrollContentContainer size={screenSize}>
          <ScrollScrollView>
            <PageContainer layout={layout} margintop={marginTop}>
              {this.renderHeaderAndInfo(schema)}
            </PageContainer>
            <PageContainer id="dynamicFormWrapper">
              <DynamicForm
                initialValues={initialValues}
                schema={schema}
                components={components}
                onSubmit={this.withValidation(onSubmit)}
                submit={submit}
                submitOnEnter={submitOnEnter}
              />
            </PageContainer>
          </ScrollScrollView>
          {multiSwitchError}
          {!omitSubmit && (
            <Fragment>
              <ScrollButtonBar style={buttonStyles}>
                <Button
                  type={buttonType}
                  right={buttonIcon}
                  middle={buttonText}
                  onClick={submit}
                />
                {extraButtonBarButtons && extraButtonBarButtons}
              </ScrollButtonBar>
              <ScrollButtonBar style={buttonStyles}>
                {extraButtons && extraButtons}
              </ScrollButtonBar>
            </Fragment>
          )}
        </ScrollContentContainer>
      );
    }

    return null;
  }
}

export const FormPage = props => (
  <BaseFormPage {...props}>
    {isWeb ? (
      <form>{props.children}</form>
    ) : (
      <Fragment>{props.children}</Fragment>
    )}
  </BaseFormPage>
);

FormPage.defaultProps = {
  omitSubmit: false,
};

FormPage.propTypes = {
  hasErrors: bool,
  schema: object.isRequired,
  omitSubmit: bool,
  userType: string,
  multiSwitchErrorText: string,
  extraButtonBarButtons: object,
  extraButtons: object,
  submitOnEnter: bool,
  translations: shape({
    validationErrors: shape({
      onlyLetters: string.isRequired,
      required: string.isRequired,
      selectAtLeastOne: string.isRequired,
      invalidEmail: string.isRequired,
      emailTaken: string.isRequired,
      onlyNumbers: string.isRequired,
      invalidNumber: string.isRequired,
      invalidPassword: string.isRequired,
      passwordsDontMatch: string.isRequired,
      invalidPhoneNumber: string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  multiSwitchErrorText: state.locale.translations.common.multiSwitchError,
  translations: {
    validationErrors: state.locale.translations.common.validationErrors,
  },
});

export default connect(
  mapStateToProps,
  null,
)(FormPage);
