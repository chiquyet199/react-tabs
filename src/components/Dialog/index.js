import React from "react";
import {Modal} from "react-native";
import {DialogOverlay} from "./Dialog.styles";
import Dialog from "./Dialog";
import {string, func, bool, object} from "prop-types";
import {themed} from "../../utils/common";

export const SimpleDialog = props => {
  const {isOpen, theme, onDialogCancel} = props;

  return (
    isOpen && (
      <Modal animationType="fade" transparent onRequestClose={onDialogCancel}>
        <DialogOverlay theme={theme}>
          <Dialog {...props} />
        </DialogOverlay>
      </Modal>
    )
  );
};

SimpleDialog.propTypes = {
  title: string,
  paragraph: string,
  confirmButtonType: string,
  cancelButtonText: string,
  confirmButtonText: string,
  onDialogCancel: func,
  onDialogConfirm: func,
  onClose: func,
  isOpen: bool.isRequired,
  theme: object.isRequired,
  globaltheme: string.isRequired,
  fullWidthText: bool,
};

export default themed(SimpleDialog);
