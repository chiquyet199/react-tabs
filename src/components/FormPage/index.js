import {connect} from "react-redux";
import {submit} from "redux-form";
import FormPage from "./FormPage";
import {go} from "../../actions/navigation";

export const mapStateToProps = state => ({
  screenSize: state.dimensions.size,
});

const mapDispatchToProps = {
  submit: () => dispatch => dispatch(submit("DynamicForm")),
  go,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FormPage);
