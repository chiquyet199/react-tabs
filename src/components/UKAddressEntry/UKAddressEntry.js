import React, {Component} from "react";
import FormPage from "../FormPage";
import {
  UKAddressEntrySchemaFragmentGenerator,
  AddressEntryStateEnum,
} from "../../utils/SchemaFragmentGenerators/SchemaFragmentGenerators";
import {func, shape, string} from "prop-types";
import Dialog from "../Dialog";
import {formatAddress} from "../../utils/string";

export default class UKAddressEntry extends Component {
  constructor(props) {
    super(props);
    this.state = {
      addressEntry: AddressEntryStateEnum.POSTCODE,
      addressToConfirm: "",
      confirmedAddress: "",
    };
  }

  handleOnAddressSelected = addressToConfirm => {
    const {onAddressSelected} = this.props;
    if (onAddressSelected) {
      onAddressSelected(addressToConfirm);
    }

    this.setState({
      addressToConfirm,
    });
  };

  handleOnAddressNotListed = () => {
    this.setState({
      addressEntry: AddressEntryStateEnum.MANUAL,
    });
  };

  handleDialogConfirm = () => {
    this.setState(
      {
        confirmedAddress: this.state.addressToConfirm,
      },
      () => {
        this.props.save(this.state.confirmedAddress);
      },
    );
  };

  handleDialogCancel = () => {
    this.setState({
      addressToConfirm: "",
    });
  };

  render() {
    const {
      globaltheme,
      save,
      translations,
      fieldName = "postcode",
      header,
      infoManual,
      infoPostcode,
    } = this.props;
    const schema = UKAddressEntrySchemaFragmentGenerator(
      this.state.addressEntry,
      this.handleOnAddressSelected,
      this.handleOnAddressNotListed,
      header,
      infoManual,
      infoPostcode,
      fieldName,
    );

    const omitSubmit =
      this.state.addressEntry === AddressEntryStateEnum.POSTCODE;

    return (
      <React.Fragment>
        <Dialog
          cancelButtonText={translations.chooseAnother}
          confirmButtonText={translations.confirm}
          globaltheme={globaltheme}
          onClose={this.handleDialogCancel}
          onDialogConfirm={this.handleDialogConfirm}
          onDialogCancel={this.handleDialogCancel}
          paragraph={formatAddress(this.state.addressToConfirm, true)}
          isOpen={this.state.addressToConfirm !== ""}
        />
        <FormPage schema={schema} omitSubmit={omitSubmit} onSubmit={save} />
      </React.Fragment>
    );
  }
}

UKAddressEntry.propTypes = {
  globaltheme: string.isRequired,
  onAddressSelected: func,
  save: func.isRequired,
  header: string,
  infoManual: string,
  infoPostcode: string,
  fieldName: string,
  translations: shape({
    confirm: string,
    chooseAnother: string,
  }),
};
