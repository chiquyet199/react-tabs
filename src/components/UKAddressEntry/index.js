import UKAddressEntry from "./UKAddressEntry";
import {connect} from "react-redux";

const mapStateToProps = state => ({
  translations: state.locale.translations.common.UKAddressEntry,
  globaltheme: state.globaltheme,
});

export default connect(
  mapStateToProps,
  {},
)(UKAddressEntry);
