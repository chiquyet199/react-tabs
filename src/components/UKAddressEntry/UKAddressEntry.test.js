import React from "react";
import {shallow} from "enzyme";
import {Platform} from "react-native";
import UKAddressEntry from "./UKAddressEntry";
import FormPage from "../FormPage";
import AddressSchema from "../../schemas/UKAddressSchema";
import Dialog from "../Dialog";
import _ from "underscore";

describe("UKAddressEntry tests", () => {
  it("renders proper snapshot - " + Platform.OS, () => {
    const component = getComponent();
    expect(component).toMatchSnapshot();
  });

  it("renders Postcode entry by default", () => {
    const form = getComponent().find(FormPage);
    const schemaProps = form.find(FormPage).prop("schema");
    expect(form).toHaveProp("schema");
    expect(schemaProps).toEqual(
      schemaProps,
      expect.objectContaining(AddressSchema.ukPostcodeEntry().properties),
    );
  });

  it("switches to manual entry when onAddressNotListed is invoked", () => {
    const component = getComponent();
    const form = component.find(FormPage);
    const schemaProps = form.find(FormPage).prop("schema");
    form.prop("schema").properties.postcode.onAddressNotListed();
    component.update();
    expect(form).toHaveProp("schema");
    expect(schemaProps).toEqual(
      schemaProps,
      expect.objectContaining(AddressSchema.ukManualEntry().properties),
    );
  });

  it("disables submit button when in postcode entry", () => {
    const form = getComponent().find(FormPage);
    expect(form).toHaveProp("omitSubmit", true);
  });

  it("enables submit button when in manual entry", () => {
    const component = getComponent();
    const form = component.find(FormPage);
    form.prop("schema").properties.postcode.onAddressNotListed();
    component.update();
    expect(component.find(FormPage)).toHaveProp("omitSubmit", false);
  });

  it("passes address selected property to FormPage schema", () => {
    const onAddressSelected = jest.fn();
    const schema = getComponent({onAddressSelected})
      .find(FormPage)
      .prop("schema");
    const schemaAddressSelected = schema.properties.postcode.onAddressSelected;
    const address = {
      line1: "1 High Street",
      line2: "Stillington",
      town: "York",
    };
    schemaAddressSelected(address);
    expect(onAddressSelected).toHaveBeenCalledWith(address);
  });

  it("passes save property to FormPage", () => {
    const save = () => {};
    const form = getComponent({save}).find(FormPage);
    expect(form).toHaveProp("onSubmit", save);
  });

  it("doesn't display confirm dialog by default", () => {
    expect(getComponent().find(Dialog)).toHaveProp("isOpen", false);
  });

  it("displays confirm dialog when address selected", () => {
    expect(getComponentWithDialog().find(Dialog)).toHaveProp("isOpen", true);
  });

  it("dialog confirm submits form", () => {
    const save = jest.fn();
    const address = "1 High Street, Stillington, York, YO60 7PD";
    const component = getComponentWithDialog(getComponent({save}), address);
    const dialog = component.find(Dialog);
    dialog.prop("onDialogConfirm")();
    component.update();
    expect(save).toHaveBeenCalledWith(address);
  });

  it("dialog cancel hides dialog", () => {
    const component = getComponentWithDialog();
    component.find(Dialog).prop("onDialogCancel")();
    component.update();
    expect(component.find(Dialog)).toHaveProp("isOpen", false);
  });

  it("dialog close hides dialog", () => {
    const component = getComponentWithDialog();
    component.find(Dialog).prop("onClose")();
    component.update();
    expect(component.find(Dialog)).toHaveProp("isOpen", false);
  });

  it("passes address to dialog", () => {
    const address = {
      line1: "1 High Street",
      line2: "Stillington",
      town: "York",
      postCode: "YO60 7PD",
    };

    expect(getComponentWithDialog(undefined, address).find(Dialog)).toHaveProp(
      "paragraph",
      "1\u00A0High\u00A0Street, Stillington, York, YO60\u00A07PD",
    );
  });

  it("includes header in schema", () => {
    const header = "header";
    const component = getComponent({header});
    const schema = component.find(FormPage).prop("schema");
    expect(schema.header).toBe(header);
  });

  it("includes infoManual and infoPostcode in schema if provided", () => {
    const infoManual = "infoManual";
    const infoPostcode = "infoPostcode";
    const component = getComponent({
      infoManual,
      infoPostcode,
    });
    const form = component.find(FormPage);
    const schema = form.prop("schema");
    expect(schema.info).toBe(infoPostcode);
    form.prop("schema").properties.postcode.onAddressNotListed();
    component.update();
    const schemaManual = component.find(FormPage).prop("schema");
    expect(schemaManual.info).toBe(infoManual);
  });

  it("use `postcode` key in schema if no fieldName is provided", () => {
    const component = getComponent();
    const schema = component.find(FormPage).prop("schema");
    const [schemaField] = Object.keys(schema.properties);
    expect(schemaField).toBe("postcode");
  });

  it("includes dynamic fieldName in schema as a key", () => {
    const fieldName = "customName";
    const component = getComponent({fieldName});
    const schema = component.find(FormPage).prop("schema");
    const [schemaField] = Object.keys(schema.properties);
    expect(schemaField).toBe(fieldName);
  });

  const getComponentWithDialog = (
    component = getComponent(),
    address = "address",
  ) => {
    const {onAddressSelected} = component
      .find(FormPage)
      .prop("schema").properties.postcode;
    onAddressSelected(address);
    component.update();

    return component;
  };

  const getComponent = props => {
    const parsedProps = _.extend(
      {
        globaltheme: "",
        onAddressSelected: null,
        save: () => {},
        translations: {
          confirm: "confirm",
          chooseAnother: "chooseAnother",
        },
      },
      props,
    );

    return shallow(<UKAddressEntry {...parsedProps} />);
  };
});
