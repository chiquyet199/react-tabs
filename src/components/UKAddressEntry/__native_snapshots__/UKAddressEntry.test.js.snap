// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UKAddressEntry tests renders proper snapshot - ios 1`] = `
<Fragment>
  <Connect(WithTheme(SimpleDialog))
    cancelButtonText="chooseAnother"
    confirmButtonText="confirm"
    globaltheme=""
    isOpen={false}
    onClose={[Function]}
    onDialogCancel={[Function]}
    onDialogConfirm={[Function]}
    paragraph=""
  />
  <Connect(Connect(FormPage))
    omitSubmit={true}
    onSubmit={[Function]}
    schema={
      Object {
        "header": undefined,
        "info": undefined,
        "properties": Object {
          "postcode": Object {
            "onAddressNotListed": [Function],
            "onAddressSelected": [Function],
            "title": "Postcode",
            "type": "string",
            "uiComponent": "postcode",
          },
        },
        "type": "object",
      }
    }
  />
</Fragment>
`;
